### ISSUE TITLE
`transpose` doing conjugation for complex input by default

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#45063]
## 🚀 Feature
As per title

## Motivation
Allows not to think too much whether input is complex or real for most linear algebra algorithms.
Eliminates the necessity to modify backward for some functions.

**Comment by ezyang** [comment]
We can (and should) add a hermitian operator; but I'm pretty unconvinced that we should silently assume the user meant H when they say T, in contravention of mathematics.

**Comment by ezyang** [comment]
cc @boeddeker

**Comment by boeddeker** [comment]
I think most people know numpy.
In numpy the transpose function does only transpose (Beside doing slightly different things).

When reading the literature, many people say "conjugate transpose" (e.g. [1]), so implementing the `transpose` operation to do also a conjugate, it would lead to confusion.

A hermitian operator could help to reduce the overhead to zero for complex gradients, when they are used for real data.
But, I am not sure, how relevant this overhead is.

[1] https://en.wikipedia.org/wiki/Conjugate_transpose

**Comment by nikitaved** [comment]
Tensorflow allows conjugation in `transpose`, although it does not do that by default.
In inner-product spaces over reals `T` is essentially an adjoint operator, so is `H` over complex numbers.
Then, what do you think about a method like `adjoint`? In most backward implementations `T` exactly is adjoint.

**Comment by anjali411** [comment]
> I think most people know numpy.
> In numpy the transpose function does only transpose (Beside doing slightly different things).
> 
> When reading the literature, many people say "conjugate transpose" (e.g. [1]), so implementing the `transpose` operation to do also a conjugate, it would lead to confusion.
I agree with @boeddeker here. I think we should follow the numpy semantics and only perform conjugate transpose when user calls `tensor.H`.

As Ed pointed out, I think tensorflow's behavior to return conjugate transpose for `torch.transpose` is unintuitive mathematically.

**Comment by pearu** [comment]
`transpose` (or `T`, in short) should do transpose only.

For conjugate transpose, I suggest introducing the method `transjugate` (or `H`, in short) as in https://en.wikipedia.org/wiki/Conjugate_transpose

Adjoint is specific to square matrices.

**Comment by t-vi** [comment]
I'll weep if you call it transjugate :cry:, it seems to be completely made up (i.e. somebody put it in wikipedia and everyone copied it from there because it reeks remotely witty): Googling "transjugate" gives about 1.480 entries, "adjoint operator" (with quotation marks) about 482000 (and for operators we can pretty much exclude that it means adjungated).

It is a bit unfortunate that the the adjugate matrix used, historically, to be called adjoint at times ("classical adjoint" as wikipedia puts it in the page linked from the mention that the adjoint matrix should not be confused with the adjugate matrix) and apparently is still taught as that in basic linear algebra. This might also explain the confusion over adjoint only bing for square matrices, which isn't the case (looking at the wikipedia page you linked, where most of the examples for adjoint matrices are, in fact, non-square).

For people thinking about matrices as representing operators - it would seem that that has been the source of much success of using adjoint matrices, including in the numerics we cater for here - it would seem more natural that _adjoint_ operators are represented by _adjoint_ matrices rather than _transjugate_ matrices. In that theme, I have never heard of _tranjugated operators_ or _self-transjugated_ anything, while _self-adjoint_ seems to be another word for Hermitian, i.e. the _adjoint_ is equal to _self_.

**Comment by boeddeker** [comment]
My favorite is `hermitian` and `.H`. 
 - `.H` is often used in formulas and would be the counterpart to `.T`.
 - `hermitian` starts with the same letter as `.H`. To it is easy to see the connection.
 - `transjugate` (Beside @t-vi comment): The name is to close to transpose.
 - `adjoint` sound very similar to adjungate.

**Comment by IvanYashchuk** [comment]
In Julia, the conjugate transposition operation is called [adjoint](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#Base.adjoint) and it's not specific for square matrices.

**Comment by nikitaved** [comment]
If `Hermitian` is adopted, then what do you do with other methods, like `symeig`? It does support Hermitian input as of now, but it is not obvious given its name.

**Comment by t-vi** [comment]
@nikitaved numpy uses numpy.linalg.eigh

**Comment by nikitaved** [comment]
Exactly, I would not expect `symeig` work for complex inputs, unless they are self-adjoint. But if `symeig` stays for Hermitian inputs, then why not make `transpose` support conjugation? ;)

**Comment by boeddeker** [comment]
> then why not make transpose support conjugation

I think support would be ok, but not by default.
So when translating `conjugate transpose` to code as `.transpose(..., conj=True)`, it is obvious, what it does (but maybe not really better than `.transpose(...).conj()`)

**Comment by nikitaved** [comment]
`conj` is no-op for non-complex input, so, probably, it is fine to just use `conj` and be happy with it.

**Comment by t-vi** [comment]
I think .adjunct() is much more concise than .transpose(conj=True). Also boolean flags are a bit of  a liability if you think of the C++ API ("boolean blindness" is the keyword.)
Having a single command one saves tensor creation. Although it's just a view, the overhead can be significant. People do lectures in which they claim Python is 1000x to slow after they measure tensor creation overhead.

**Comment by pearu** [comment]
From a general perspective, `.transpose(..., conj=True)` indicates that a matrix operation (`transpose` in this case) can be attributed with an element-wise unary operation (`conjugate` in this case).
Using the same pattern, one could have `.transpose(.., negate=True)`, etc.
I wonder if people would be willing to accept and extend this pattern?

While `transjugate` is not widely used in terms of google hits, the term has a unique meaning and has been documented by independent sources (google: transjugate) to mean conjugate transpose operation.

**Comment by t-vi** [comment]
Just for some stats: There are about 35 articles in google scholar from about 10 distinct author's that call it transjugate. About 8 of them will have gotten the term off Wikipedia because it's outside their language or field. One dates from the 1960s so we can be assured that it was indeed a uniquely unique term until 16 December 2007.  In google books there are about 23 hits for transjugate, but after 19 or so (considerably overlapping with scholar hits) the snippets are more about "preventing transjugation of the Afridi nation", which might not necessarily relate to the adjoint matrix as much.
The reason 16 December 2007 is important to the history of the term transjugate, is because Wikipedia used to spell it tranjugate (without the s) until that time and in addition to the 1500 web pages copying transjugate off Wikipedia, there are 1100 calling it tranjugate, likely equally inspired by Wikipedia.
The term transjugate is so obscure Wikipedia doesn't even have a redirect from transjugate to the page about it.
In summary, it would seem to me is very much a fringe term that hardly anyone except the very few except people who haven't been exposed to it in mathematical context nor look at textbooks but look things up in Wikipedia instead use.
I must admit I have get no fun from any of this discussion and if you feel the urge to create an API that screams "the mathematics is foreign to us" because that is so unique, you'll just have to do what you have to do. The Julia people were apparently more mathematically minded in advocation of names than Quansight is, but I guess that's in line with the image they have, too.

**Comment by hameerabbasi** [comment]
> My favorite is `hermitian` and `.H`.

I like `hermitian`, `.H` implies an `O(1)` operation as it is a property, which this isn't. For this reason, I don't like it.

**Comment by rgommers** [comment]
> `.H` implies an `O(1)` operation as it is a property, which this isn't.

This is also the rationale for why numpy never added `.H`.

I'd go with `.hermitian()` or `.adjoint()`, both seem like sensible names.

**Comment by nikitaved** [comment]
Hermitian could be confusing as it has the meaning akin symmetry for real matrices, in my opinion.

**Comment by ezyang** [comment]
A few things:

1. If `conjugate views <https://github.com/pytorch/pytorch/pull/44799>`_ go in, hermitian will be O(1). So we can (and should) provide a `.H` attribute.
2. I don't think we, as framework developers, should really be in the business of telling mathematicians what terms are and aren't ambiguous. So if a term like "hermitian" is standard nomenclature in the mathematical community, we should support it. (I also happen to think that "taking the Hermitian of a matrix" is pretty unambiguous, since the other property is a description of a matrix, rather than an action you take on it.
3. There isn't really any reason we can't provide multiple aliases for an operation; it's not an all or nothing deal. Though, granted, this is not that Pythonic (there is one obvious way to do it)
4. "transpose" and "symeig" are not analogous situations. It is very reasonable to want to do a non-conjugate transpose on a matrix. But it makes very little sense to do a "symeig" on a symmetric but non-hermitian complex matrix, this is a very unnatural thing to do and won't give you real eigenvalues (which means you will have trouble with dtypes anyway). That being said, we should provide "eigh" (the Numpy naming) to disambiguate this situation.
5. Hermitian (operator), the verb, is not specific to square matrices (it is only the adjective, Hermitian, that is applicable to square matrices only)

**Comment by anjali411** [comment]
From the discussion, it seems like the general consensus is to call the conjugate transpose operator `.hermitian()` in PyTorch. If we end up adding [conjugate views](https://github.com/pytorch/pytorch/pull/44799), I think we should add `tensor.H` as a follow-up which would be a view (O(1) operation).

**Comment by Lezcano** [comment]
My two cents on this:

I completely agree with the `X.H` attribute and the `O(1)` operation discussion.

On the naming of the full operation, I prefer `adjoint` over `hermitian` for the following reasons:

I think that "taking the Hermitian of a matrix" is somewhat an odd thing to say. I have personally never encountered it in maths writings. If I read the code `X.hermitian()`, I would think that it's an operation that  it is equivalent to the code `X == X.H`.

In this sense, the term adjoint is more general and offers no confusion. It is clear that, in the real, case, it refers to $X^T$ and on the complex case it refers to $X^H$. It is also pretty common to see the phrase "taking adjoints" in mathematical papers. At the end of the day, all the `backward` methods do is to compute the adjoint of the linearisation (differential) of each map, hence the name of the backprop algorithm in mathematics: [The adjoint method](https://en.wikipedia.org/wiki/Adjoint_state_method).

**Comment by anjali411** [comment]
> I think that "taking the Hermitian of a matrix" is a somewhat odd thing to say. I have personally never encountered it in maths writings. If I read `X.hermitian()`, I would think that it's a boolean operation that is equivalent to the code `X == X.H`.

sold. `adjoint` and `conjugate_transpose` both seem like reasonable options however I prefer the first over the other since it's less wordy.

**Comment by Lezcano** [comment]
I'm going to add an `def adjoint(dim0=-2, dim1=-1)` operator to mean `.transpose(-2, -1).conj()`.

The next interesting point is that we'd like to have a property similar to `T` that computes `H`. Now, `T` is completely broken for multidimensional tensors, which is annoying. This opens a number of possibilities:
- Add `H` to mean `tensor.T.conj()`. In this case we would be deliberately adding another broken function.
- Add `H` with the right behaviour and start the process of deprecating `T` when called on a tensor with more than two dims. I think this should not be too painful given that who uses the behaviour of `T` on a tensor??
- Add `H` with the wrong behaviour, and then add `mT` and `mH` (short for matrix Transpose) that do the right thing and encourage people to use these. This was proposed by @rgommers , who said that it was proposed for NumPy but no one has ever got around to adding it.

A note in option 2, we could direct people to use `.H` as a replacement for `.T` while `.T` is updated, as `.H` on a real tensor would do the correct thing.

I like option 2 best, but I'd be happy to go for option 3 if it seems like too much effort.

In support for why we would like shortcuts like these, I believe that they are extremely important for readability. Consider the case when we want to tell the user that solving `XA = B` is the same as solving `A^T X^T = B^T`. With the current notation we have to tell them to write
```python
linalg.solve(A.transpose(-2, -1), B.transpose(-2, -1)).transpose(-2, -1)
```
while with `T`'s or `H`'s, this would read

```python
linalg.solve(A.T, B.T).T
```

Also, the backward of matrix functions is full of definitions as `A_H = A.transpose(-2, -1).conj()` to make the formulas readable. This could be avoided by calling `A.H()` in C++.

**Comment by ezyang** [comment]
> Add H with the right behaviour and start the process of deprecating T when called on a tensor with more than two dims. I think this should not be too painful given that who uses the behaviour of T on a tensor??

This seems reasonable to me. Since H doesn't exist today it is not a big deal to have it only work on 2D tensors.

**Comment by Lezcano** [comment]
Sorry, by "right behaviour of tensor.H I meant for it to be equivalent to `tensor.transpose(-2, -1).conj()` i.e. treat tensor as a batch of matrices.

I corrected that in the previous message.

**Comment by Lezcano** [comment]
There was a recent discussion at the array API consortium. It is summarised here: https://github.com/data-apis/array-api/issues/228#issuecomment-904009162
They decided to go for `T` / `H` for matrices (i.e. we could deprecate their behaviour for tensors) and then `mT` and `mH` to do `transpose(-2, -1)` and `transpose(-2, -1).conj()`.

**Comment by ezyang** [comment]
OK, that seems reasonable.
